source/extensions/filters/http/cdn_loop/parser.cc

```diff
@@ -0,0 +1,359 @@
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected closing quote at position %d; found '%c'.", input.next(), context.peek()));
+  }
+  context.increment();
+
+  return context;
+}
+
+StatusOr<ParseContext> parseToken(const ParseContext& input) {
+  ParseContext context = input;
+  while (!context.atEnd()) {
+    const char c = context.peek();
+    // Put alphanumeric, -, and _ characters at the head of the list since
+    // they're likely to be used most often.
+    if (isAlpha(c) || isDigit(c) || c == '-' || c == '_' || c == '!' || c == '#' || c == '$' ||
+        c == '%' || c == '&' || c == '\'' || c == '*' || c == '+' || c == '.' || c == '^' ||
+        c == '`' || c == '|' || c == '~') {
+      context.increment();
+    } else {
+      break;
+    }
+  }
+  if (context.next() == input.next()) {
+    if (context.atEnd()) {
+      return absl::InvalidArgumentError(absl::StrFormat(
+          "expected token starting at position %d; found end of input", input.next()));
+    } else {
+      return absl::InvalidArgumentError(absl::StrFormat(
+          "expected token starting at position %d; found %c", input.next(), context.peek()));
+    }
+  }
+
+  return context;
+}
+
+StatusOr<ParseContext> parsePlausibleIpV6(const ParseContext& input) {
+  ParseContext context = input;
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected IPv6 literal at position %d; found end-of-input", context.next()));
+  }
+
+  if (context.peek() != '[') {
+    return absl::InvalidArgumentError(absl::StrFormat("expected opening '[' of IPv6 literal at "
+                                                      "position %d; found %c",
+                                                      context.next(), context.peek()));
+  }
+  context.increment();
+
+  while (true) {
+    if (context.atEnd()) {
+      break;
+    }
+    const char c = context.peek();
+    if (!(isHexDigitCaseInsensitive(c) || c == ':' || c == '.')) {
+      break;
+    }
+    context.increment();
+  }
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected closing ']' of IPv6 literal at position %d "
+                        "found end-of-input",
+                        context.next()));
+  }
+  if (context.peek() != ']') {
+    return absl::InvalidArgumentError(absl::StrFormat("expected closing ']' of IPv6 literal at "
+                                                      "position %d; found %c",
+                                                      context.next(), context.peek()));
+  }
+  context.increment();
+
+  return context;
+}
+
+StatusOr<ParsedCdnId> parseCdnId(const ParseContext& input) {
+  ParseContext context = input;
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected cdn-id at position %d; found end-of-input", context.next()));
+  }
+
+  if (StatusOr<ParseContext> ipv6 = parsePlausibleIpV6(context); ipv6) {
```

Comment:
Are we ever using the status here?

I love the consistency and debugability of StatusOr but I'm less excited about doing multiple string format calls when the headers are valid, and we're trying to sort out which type of valid they are.

Alternately since I think we can determine if it's plausible ipv6 or not from a one character peek, we could do the one character peek and try parsing ipv6 or token based on that?

Reply:
Using peek worked out nicely.  I did it here and in the values of parameters (where we can parse double-quoted strings).